ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -697,18 +702,89 @@ public void findEntriesSinceOnRestartTest() throws IOException, StoreException {
     // token should get reset internally, no keys should be returned and the returned token should be correct (offset in
     // it will be the current log end offset = firstRecordFileSpan.getEndOffset()).
-    StoreFindToken expectedEndToken = new StoreFindToken(firstRecordFileSpan.getEndOffset(), sessionId);
+    StoreFindToken expectedEndToken =
+        new StoreFindToken(firstRecordFileSpan.getEndOffset(), sessionId, incarnationId, true);
     expectedEndToken.setBytesRead(bytesRead);
     doFindEntriesSinceTest(startToken, Long.MAX_VALUE, Collections.EMPTY_SET, expectedEndToken);
 
     // create a token that is not past the index end offset on startup after recovery. Should work as expected
-    startToken = new StoreFindToken(lastRecordOffset, oldSessionId);
-    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId);
+    startToken = new StoreFindToken(lastRecordOffset, oldSessionId, incarnationId, false);
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
     expectedEndToken.setBytesRead(bytesRead);
     doFindEntriesSinceTest(startToken, Long.MAX_VALUE, Collections.singleton(newId), expectedEndToken);
+
+    // same test as above, but with old incarnationId
+    bytesRead = log.getDifference(firstRecordFileSpan.getEndOffset(), new Offset(log.getFirstSegment().getName(), 0));
+    // create a token that will be past the index end offset on startup after recovery with old incarnationId
+    startToken = new StoreFindToken(secondRecordFileSpan.getEndOffset(), oldSessionId, oldIncarnationId, false);
+    // token should get reset internally, returning all the keys and the returned token should be correct (offset in
+    // it will be the starting offset of last entry which is firstRecordFileSpan.getStartOffset())
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest(startToken, Long.MAX_VALUE, allKeys.keySet(), expectedEndToken);
+
+    // create a token that is not past the index end offset on startup after recovery, but with old incarnationId.
+    // again the token should be reset internally, all keys should be returned and the returned token should be correct
+    // (offset in it will be the starting offset of last entry which is firstRecordFileSpan.getStartOffset())
+    startToken = new StoreFindToken(lastRecordOffset, oldSessionId, oldIncarnationId, false);
+    expectedEndToken = new StoreFindToken(firstRecordFileSpan.getStartOffset(), sessionId, incarnationId, false);
+    expectedEndToken.setBytesRead(bytesRead);
+    doFindEntriesSinceTest(startToken, Long.MAX_VALUE, allKeys.keySet(), expectedEndToken);
+
+    // test cases where no new puts/deletes have gone in after restart.
+    // after restart, replica 1(assuming no new puts have gone in) will return a journal based token referring to
```

Comment:
don't think you need to talk about replicas here. This is all store code and there is no replication here - it would be better to talk in terms of the store only.